src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,342 @@
+{
+    if (strcasecmp(name, field) == 0) {
+        tag = value;
+        return 1;
+    }
+    return 0;
+}
+
+static int
+usf_tags_target(void * context, const char * name, const char * value)
+{
+    struct usf_tags * tags = ( struct usf_tags * ) context;
+    if (!value) return 0;
+
+    if (usf_save_tag("title", name, value, tags->title));
+    else if (usf_save_tag("artist", name, value, tags->artist));
+    else if (usf_save_tag("composer", name, value, tags->composer));
+    else if (usf_save_tag("game", name, value, tags->game));
+    else if (usf_save_tag("year", name, value, tags->year));
+    else if (usf_save_tag("copyright", name, value, tags->copyright));
+    else if (usf_save_tag("usfby", name, value, tags->usfby));
+    else if (usf_save_tag("genre", name, value, tags->genre));
+    else if (usf_save_tag("length", name, value, tags->length));
+    else if (usf_save_tag("fade", name, value, tags->fade));
+    else if (usf_save_tag("track", name, value, tags->track)){}
+    return 0;
+}
+
+static double
+get_length_from_string(std::string length)
+{
+    const char * str_length = length.c_str();
+    double ttime = 0, temp = 0, mult = 1, level = 1;
+    char Source[1024];
+    char * src = Source + strlen(str_length);
+    strcpy(&Source[1], str_length);
+    Source[0] = 0;
+
+    while(*src)
+    {
+        if((*src >= '0') && (*src <= '9'))
+        {
+            temp += ((*src - '0') * mult);
+            mult *= 10;
+        }
+        else
+        {
+            mult = 1;
+            if(*src == '.')
+            {
+                ttime = temp;
+                temp = 0;
+            }
+            else if(*src  == ':')
+            {
+                ttime += (temp * (1000 * level));
+                temp = 0;
+                level *= 60;
+            }
+        }
+        src--;
+    }
+
+    ttime += (temp * (1000 * level));
+    return ttime/1000;
+}
+
+static void
+track_lengths(std::string length, std::string fade, double& track_length, double& fade_time, double& total_length)
+{
+    track_length = -1;
+    fade_time = 0;
+
+    if (!length.empty()) {
+        if (!fade.empty()) {
+            fade_time = get_length_from_string(fade);
+        }
+        track_length = get_length_from_string(length);
+    }
+    total_length = track_length + fade_time;
+
+}
+
+static void
+usf_file_decode(Decoder &decoder, Path path_fs)
+{
+    /* Load the file */
+
+    usf_tags tags;
+    usf_loader_state state;
+    state.emu = malloc(usf_get_state_size());
+    usf_clear(state.emu);
+
+    const char* path = path_fs.c_str();
+    const int psf_version_tags = psf_load(path, &stdio_callbacks, 0, 0, 0, usf_tags_target, &tags, 0);
+    const int psf_version_state = psf_load( path, &stdio_callbacks, psf_version_tags, usf_loader, &state, usf_info, &state, 0 );
+
+    if ( psf_version_tags != 0x21 || psf_version_state <= 0 ) {
+        LogWarning(usf_domain, "Error loading usf file");
+        free(state.emu);
```

Comment:
Use `AtScopeExit()` to make freeing the buffer reliable, if you really need `malloc()`. Adding `free()` to every `return` is error prone.